File: src/main/java/strings/longest/substring/without/repeating_3/Solution.cpp
#include <iostream>
#include <cstring>
#include <unordered_map>
using namespace std;

class Solution {

public:
   int lengthOfLongestSubstring(string s) {
    int n = s.length();
    char char_array[n + 1];
    strcpy(char_array, s.c_str());
    int startIdx = 0;
    int endIdx = 0;
    int max = 0;
    unordered_map<char, int> m;
    for (int i = 0; i < n; i++) {
        char curr = char_array[i];
        if(m.find(curr) != m.end()) {
            int idx = m[curr] + 1;
            if(startIdx< idx){
                startIdx = idx;
            }
        }
        endIdx++;
        int result = endIdx - startIdx;
        if(result> max){
            max = result;
        }
        m[curr]=i;
    }
      return max;
   }
};


int main()
{
    Solution sol;
    cout << sol.lengthOfLongestSubstring("wobgrovw");
    return 0;
}

File: src/main/java/strings/longest/substring/without/repeating_3/Solution.java
package strings.longest.substring.without.repeating_3;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

class Solution {

    public int lengthOfLongestSubstring(String s) {
        int max = 0;
        Map<Character, Integer> map = new HashMap<>();
        int startIdx = 0;
        int endIdx = 0;
        for (int i = 0; i < s.length(); i++) {
            char curr = s.charAt(i);
            if (map.containsKey(curr)) {
                int idx = map.get(curr) + 1;
                if (idx > startIdx) {
                    startIdx = idx;
                }
            }
            endIdx++;
            int result = endIdx - startIdx;
            if (max < result) {
                max = result;
            }
            map.put(curr, i);
        }
        return max;
    }

    public int lengthOfLongestSubstringInitial(String s) {
        int max = 0;
        Set<Character> uniqueChars = new HashSet<>();
        for (int i = 0; i < s.length(); i++) {
            int temp = 1;
            for (int j = i + 1; j < s.length(); j++) {
                char curr = s.charAt(j);
                if (s.charAt(i) == curr || uniqueChars.contains(curr)) {
                    break;
                }
                uniqueChars.add(curr);
                temp++;
            }
            max = max < temp ? temp : max;
            uniqueChars.clear();
        }
        return max;
    }

}